dateutil/parser.py

```diff
@@ -742,318 +763,215 @@ def _parse(self, timestr, dayfirst=None, yearfirst=None, fuzzy=False,
 
                 # Check for a timezone name
-                if (res.hour is not None and len(l[i]) <= 5 and
-                        res.tzname is None and res.tzoffset is None and
-                        not [x for x in l[i] if x not in
-                             string.ascii_uppercase]):
+                elif _could_be_tzname(res.hour, res.tzname, res.tzoffset, l[i]):
                     res.tzname = l[i]
                     res.tzoffset = info.tzoffset(res.tzname)
-                    i += 1
 
                     # Check for something like GMT+3, or BRST+3. Notice
                     # that it doesn't mean "I am 3 hours after GMT", but
                     # "my time +3 is GMT". If found, we reverse the
                     # logic so that timezone parsing code will get it
                     # right.
-                    if i < len_l and l[i] in ('+', '-'):
-                        l[i] = ('+', '-')[l[i] == '+']
+                    if i+1 < len_l and l[i+1] in ('+', '-'):
+                        l[i+1] = ('+', '-')[l[i+1] == '+']
                         res.tzoffset = None
                         if info.utczone(res.tzname):
                             # With something like GMT+3, the timezone
                             # is *not* GMT.
                             res.tzname = None
 
-                    continue
 
                 # Check for a numbered timezone
-                if res.hour is not None and l[i] in ('+', '-'):
+                elif res.hour is not None and l[i] in ('+', '-'):
                     signal = (-1, 1)[l[i] == '+']
-                    i += 1
-                    len_li = len(l[i])
+                    len_li = len(l[i+1])
 
+                    # TODO: check that l[i+1] is integer?
                     if len_li == 4:
                         # -0300
-                        res.tzoffset = int(l[i][:2])*3600+int(l[i][2:])*60
-                    elif i+1 < len_l and l[i+1] == ':':
+                        hour_offset = int(l[i+1][:2])
+                        min_offset = int(l[i+1][2:])
+                    elif i+2 < len_l and l[i+2] == ':':
                         # -03:00
-                        res.tzoffset = int(l[i])*3600+int(l[i+2])*60
+                        hour_offset = int(l[i+1])
+                        min_offset = int(l[i+3]) # TODO: Check that l[i+3] is minute-like?
                         i += 2
                     elif len_li <= 2:
                         # -[0]3
-                        res.tzoffset = int(l[i][:2])*3600
+                        hour_offset = int(l[i+1][:2])
+                        min_offset = 0
                     else:
-                        return None, None
-                    i += 1
+                        raise InvalidDatetimeError(timestr)
 
-                    res.tzoffset *= signal
+                    res.tzoffset = signal * (hour_offset*3600 + min_offset*60)
 
+                    # TODO: Are we not requiring that res.tzname be None?
```

Comment:
I think that there's a question of adjudication if you find two things that have equal claim to being a time zone in the string. Right now I guess it just takes the most recent thing that it finds that matches?